src/sql/workbench/parts/objectExplorer/browser/serverTreeRenderer.ts

```diff
@@ -152,19 +153,44 @@ export class ServerTreeRenderer implements IRenderer {
 		let iconLowerCaseName = iconName.toLocaleLowerCase();
 		templateData.icon.classList.add(iconLowerCaseName);
 
+		if (treeNode.iconPath) {
+			renderIcon(templateData.icon, treeNode.iconPath);
+		}
+
 		templateData.label.textContent = treeNode.label;
 		templateData.root.title = treeNode.label;
 	}
 
+	private getIconPath(connection: ConnectionProfile): { light: string, dark: string } {
+		let iconPath = undefined;
+		let serverInfo = this._connectionManagementService.getServerInfo(connection.id);
+		let providerProperties = this._connectionManagementService.getProviderProperties(connection.providerName);
+		if (serverInfo && providerProperties) {
+			try {
+				let icons: { condition: string, path: object }[] = providerProperties['iconPath'];
+				if (icons) {
+					iconPath = icons.find(e => {
+						let isChosen = new Function('serverInfo', `return ${e.condition}`);
```

Comment:
2 things:
1. We don't define dynamic functions anywhere in the code. This has a bad code smell to me - why would we do this? You could do very arbitrary things with a malicious package.json contribution since it allows an extension to inject arbitrary javascript at this point. 
2. We use `ContextKeyExpr.deserialize(item.when)` when registering contributions (e.g. this should be done in `connectionProviderExtension.ts`), then we create a scoped context (inject an IContextKeyService, then use createScoped and set the serverinfo as a key), then finally use `contextKeyService.contextMatchesRule` to compare.
  - See viewExtensionPoint.ts line 384 for use of deserialize, and menuService.ts line 113 for use of `contextKeyService.contextMatchesRule`.

Reply:
fixed.